lib/simple-seeder.js

```diff
@@ -0,0 +1,221 @@
+const Hyperbee = require('hyperbee')
+const Hyperdrive = require('hyperdrive')
+const Seeders = require('@hyperswarm/seeders')
+const speedometer = require('speedometer')
+const HypercoreId = require('hypercore-id-encoding')
+const mutexify = require('mutexify/promise')
+
+module.exports = class SimpleSeeder {
+  constructor (store, swarm, opts = {}) {
+    this.store = store
+    this.swarm = swarm
+
+    this.backup = opts.backup === true
+
+    this.lock = mutexify()
+    this.resources = new Map()
+
+    this.destroying = false
+  }
+
+  async add (key, opts) {
+    const release = await this.lock()
+    try {
+      return await this._add(key, opts)
+    } finally {
+      release()
+    }
+  }
+
+  async _add (key, opts) {
+    if (this.destroying) return
+    if (this.has(key)) throw new Error('Resource already added')
+
+    const publicKey = HypercoreId.decode(key)
+    const id = HypercoreId.encode(publicKey)
+
+    // Maybe all this belongs to a "Resource" class, including the internals _createCore/Bee/etc
+    const info = {
+      key: id,
+      type: opts.type,
+      instance: null,
+      description: opts.description || null,
+      seeders: null,
+      discovery: null,
+      blocks: { down: speedometer(), up: speedometer() },
+      network: { down: speedometer(), up: speedometer() },
+      external: opts.external !== false
+    }
+
+    if (opts.type === 'core') info.instance = await this._createCore(publicKey, info)
+    else if (opts.type === 'bee') info.instance = await this._createBee(publicKey, info)
+    else if (opts.type === 'drive') info.instance = await this._createDrive(publicKey, info, opts)
+    else if (opts.type === 'list') info.instance = await this._createBee(publicKey, info, { keyEncoding: 'utf-8', valueEncoding: 'json' })
+    else if (opts.type === 'seeders') {
+      info.type = 'drive'
+      info.instance = await this._createDrive(publicKey, info, { seeders: true })
+    } else throw new Error('Type is required')
+
+    this.resources.set(id, info)
+
+    return info
+  }
+
+  async put (key, opts) {
+    const release = await this.lock()
+    try {
+      return await this._put(key, opts)
+    } finally {
+      release()
+    }
+  }
+
+  async _put (key, opts) {
+    if (!this.has(key)) return this._add(key, opts)
+
+    const info = this.get(key)
+
+    if (info.type !== opts.type) {
```

Comment:
How does this case happen? Seems strange to have a key be added as one type, and then later added again as another type

Reply:
The commonest scenario would probably be promoting a core to a bee (and vice versa):
`await tracker.put(key, { type: 'core' })`
And later
`await tracker.put(key, { type: 'bee' })`

I guess someone could add `drive.core` and `drive.blobs` as isolated bees

Unsure if there is any other scenario but the point is that it can happen so we can easily handle it